dhall/src/Dhall.hs

```diff
@@ -900,6 +909,103 @@ instance (Interpret a, Interpret b) => Interpret (a, b)
 auto :: Interpret a => Type a
 auto = autoWith defaultInterpretOptions
 
+newtype DhallFix f t = DhallFix { unDhallFix :: t }
+newtype FixVar = FixVar { unFixVar :: (Expr Src X) }
+
+-- This is a helper instance used from implementation of Interpret for DhallFix
+instance Interpret FixVar where
+    autoWith _ = Type {..}
+        where
+        extract (App _ e0) = pure (FixVar e0)
+        extract e          = typeError expected e
+
+        expected = "t"
+
+type InterpretFix f t =
+   ( Traversable f
+   , Corecursive t
+   , Interpret (f FixVar)
+   , Base t ~ f
+   )
+
+instance InterpretFix f t => Interpret (DhallFix f t) where
+    autoWith opts = Type { extract = extr, expected = expe }
+        where
+        extr (Lam _ _ (Lam _ _ e)) = DhallFix <$> buildF e
+        extr e                     = typeError expe e
+
+        -- buildF = elgot f g <=< extract (auto @FixVar)
+        viaMonadic a b x = fromMonadic $ toMonadic . a =<< toMonadic (b x)
```

Comment:
Minor suggestion: this would be slightly more symmetric (and closer to the comment right above) if written as:

```haskell
viaMonadic f g = fromMonadic . (toMonadic . f <=< toMonadic . g)
```